process/block/preprocess/transactions.go

```diff
@@ -630,80 +876,127 @@ func (txs *transactions) createAndProcessMiniBlock(
 			txs.gasHandler.RemoveGasConsumed([][]byte{txHash})
 			txs.gasHandler.RemoveGasRefunded([][]byte{txHash})
 
-			gasConsumedByMiniBlockInSenderShard = oldGasConsumedByMiniBlockInSenderShard
-			gasConsumedByMiniBlockInReceiverShard = oldGasConsumedByMiniBlockInReceiverShard
-
+			gasConsumedByMiniBlocksInSenderShard = oldGasConsumedByMiniBlocksInSenderShard
+			mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] = oldGasConsumedByMiniBlockInReceiverShard
+			totalGasConsumedInSelfShard = oldTotalGasConsumedInSelfShard
 			continue
 		}
 
-		gasRefunded := txs.gasHandler.GasRefunded(txHash)
-		gasConsumedByMiniBlockInReceiverShard -= gasRefunded
-		if senderShardId == receiverShardId {
-			gasConsumedByMiniBlockInSenderShard -= gasRefunded
-		}
+		senderAddressToSkip = []byte("")
 
-		if !errors.Is(err, process.ErrFailedTransaction) {
-			miniBlock.TxHashes = append(miniBlock.TxHashes, txHash)
+		gasRefunded := txs.gasHandler.GasRefunded(txHash)
+		mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] -= gasRefunded
+		if senderShardID == receiverShardID {
+			gasConsumedByMiniBlocksInSenderShard -= gasRefunded
+			totalGasConsumedInSelfShard -= gasRefunded
 		}
-		addedTxs++
 
-		if addedTxs >= spaceRemained { // max transactions count in one block was reached
-			log.Debug("max txs accepted in one block is reached",
-				"num added txs", len(miniBlock.TxHashes),
-				"total txs", len(orderedTxs),
-			)
+		if errors.Is(err, process.ErrFailedTransaction) {
+			if !firstInvalidTxFound {
+				firstInvalidTxFound = true
+				txs.blockSizeComputation.AddNumMiniBlocks(1)
+			}
 
-			log.Debug("mini block info",
-				"gas consumed in sender shard", gasConsumedByMiniBlockInSenderShard,
-				"gas consumed in receiver shard", gasConsumedByMiniBlockInReceiverShard,
-				"gas consumed in self shard", txs.gasHandler.TotalGasConsumed(),
-				"txs ordered", len(orderedTxs),
-				"txs added", len(miniBlock.TxHashes))
+			txs.blockSizeComputation.AddNumTxs(1)
+			continue
+		}
 
-			return miniBlock, nil
+		if _, ok := mapMiniBlocks[receiverShardID]; !ok {
+			mapMiniBlocks[receiverShardID] = txs.createEmptyMiniBlock(senderShardID, receiverShardID, block.TxBlock)
+			txs.blockSizeComputation.AddNumMiniBlocks(1)
 		}
+
+		mapMiniBlocks[receiverShardID].TxHashes = append(mapMiniBlocks[receiverShardID].TxHashes, txHash)
+		txs.blockSizeComputation.AddNumTxs(1)
+		numTxsAdded++
 	}
 
-	if addedTxs > 0 {
+	miniBlocks := txs.getMiniBlockSliceFromMap(mapMiniBlocks)
+
+	for _, miniBlock := range miniBlocks {
 		log.Debug("mini block info",
-			"gas consumed in sender shard", gasConsumedByMiniBlockInSenderShard,
-			"gas consumed in receiver shard", gasConsumedByMiniBlockInReceiverShard,
-			"gas consumed in self shard", txs.gasHandler.TotalGasConsumed(),
-			"txs ordered", len(orderedTxs),
+			"sender shard", miniBlock.SenderShardID,
+			"gas consumed", gasConsumedByMiniBlocksInSenderShard,
```

Comment:
same gasConsumed is printed multiple times, should be only once.

Reply:
Done